Multimap-As-A-First-Class-Data-Structure.asciidoc

```diff
@@ -0,0 +1,125 @@
+= Dynamic RBAC
+
+== Tracking Issue
+
+https://issues.redhat.com/browse/ISPN-13853
+
+== Description
+
+Multimap is a data structure that has gotten the attention from Infinispan users.
+The structure today lacks evolutions and support of features and new API of Infinispan:
+
+* REST API support
+* Supporting duplicates
+* Reactive API integration and development
+* Integration with Quarkus through the extension
+* Search and query support
+
+== Compatibility impact
+
+Ensure that the old Remote API and configuration still works.
+
+Under the hood, we will still keep using distributed caches, but we will know with the configuration that those
+caches are meant to behave as multimaps. This way we will be able to change the implementation details in
+the future if the structure becomes even more popular.
+
+Indexing and Query should evolve to detect that the content to be indexed is inside the list/set of the key/value.
+Query will be supported *only* if indexing is enabled.
+
+== Security impact
+
+N/A
+
+== Configuration Schema
+
+[source,xml]
+----
+<xs:complexType name="multimap">
+    <xs:complexContent>
+      <xs:extension base="tns:distributed-cache">
+        <xs:attribute name="duplicates" type="xs:boolean" default="false">
+          <xs:annotation>
+            <xs:documentation>Determines if the multimap supports duplicates in the values. True or false (default).</xs:documentation>
+          </xs:annotation>
+        </xs:attribute>
+      </xs:extension>
+    </xs:complexContent>
+</xs:complexType>
+----
+
+
+== Public API
+
+The work will be done for the new API only.
+
+== Deprecations
+
+* `org.infinispan.client.hotrod.multimap.RemoteMultimapCache`
+* `org.infinispan.client.hotrod.multimap.RemoteMultimapCacheManager`
+* `org.infinispan.client.hotrod.multimap.RemoteMultimapCacheManagerFactory`
+* `org.infinispan.client.hotrod.multimap.MultimapCacheManager`
+* `org.infinispan.client.hotrod.multimap.RemoteMultimapCache`
+* `org.infinispan.client.hotrod.multimap.MetadataCollection`
+
+== Hot Rod API
+
+The new API will be developed. Provide implementations for the available APIs:
+
+* org.infinispan.api.async.SyncMultiMap
+* org.infinispan.api.async.SyncMultiMaps
+* org.infinispan.api.async.MutinyMultiMap
+* org.infinispan.api.async.MutinyMultiMaps
+* org.infinispan.api.async.AsyncMultiMap
+* org.infinispan.api.async.AsyncMultiMaps
+
+== REST API
+
+Multimap REST API should be the almost the same as the Cache one.
+The difference is in the endpoint.
+
+Examples:
+[source]
+----
+GET /rest/v2/multimaps/
+
+GET /rest/v2/multimaps/{multimapName}?action=???
+
+DELETE /rest/v2/multimaps/{multimapName}/{multimapKey}
+----
+
+* CRUD multimap (providing a configuration to create a multimap, templates not supported)
+* Display Multimap content entries
+* Get values by key
```

Comment:
We also need an equivalent of the `CacheEntry<K, V> Cache.getEntry()` call. Unfortunately this needs to have a different interface per type

```java
// sync
SyncCacheEntryCollection<K, V> MultiMap.getEntry(K key, CacheOptions options);

public interface SyncCacheEntryCollection<K, V> {
   K key();

   CloseableIterable<V> values();

   CacheEntryMetadata metadata();
}
// async
CompletionStage<AsyncCacheEntryCollection<K, V>> MultiMap.getEntry(K key, CacheOptions options);

public interface AsyncCacheEntryCollection<K, V> {
   K key();

   Flow.Publisher<V> values();

   CacheEntryMetadata metadata();
}
// mutiny
Uni<MutinyCacheEntryCollection<K, V>> MultiMap.getEntry(K key, CacheOptions options);

public interface MutinyCacheEntryCollection<K, V> {
   K key();

   Multi<V> values();

   CacheEntryMetadata metadata();
}
```

Reply:
This also brings up the question do values each have their own metadata such as expiration? The API as proposed here wouldn't allow that.

Reply:
For now not. Do we want that?

Reply:
@tristantarrant @wburns how would we include that ?